x/valset-pref/validator_set.go

```diff
@@ -17,19 +16,61 @@ type valSet struct {
 	amount  sdk.Dec
 }
 
-func (k Keeper) SetValidatorSetPreference(ctx sdk.Context, delegator string, preferences []types.ValidatorPreference) ([]types.ValidatorPreference, error) {
-	// check if a user already has a validator-set created
-	existingValidators, found := k.GetValidatorSetPreference(ctx, delegator)
-	if found {
-		// check if the new preferences is the same as the existing preferences
-		isEqual := k.IsValidatorSetEqual(preferences, existingValidators.Preferences)
-		if isEqual {
-			return nil, fmt.Errorf("The preferences (validator and weights) are the same")
+// SetValidatorSetPreference creates or updates delegators validator set.
+// 1. if valset doesnot exist & user doesnot have existing delegation, store new valset.
+// 2. if valset doesnot exist, but user has existing delegation, convert it to valset and store it.
+// 3. if valset exists, but user doesnot have existing delegation, modify current valset.
+// 4. if valset exists & user has existing delegation, modify current valset.
+func (k Keeper) SetValidatorSetPreference(ctx sdk.Context, delegator string, preferences []types.ValidatorPreference) (types.ValidatorSetPreferences, error) {
+	existingValSet, found := k.GetValidatorSetPreference(ctx, delegator)
+	if !found {
+		delAddr, err := sdk.AccAddressFromBech32(delegator)
+		if err != nil {
+			return types.ValidatorSetPreferences{}, err
+		}
+
+		// check if there is existing delegations
+		existingDelegations, err := k.GetExistingStakingDelegations(ctx, delAddr)
+		if err != nil {
+			// check if the new preferences are valid
+			validPreference, err := k.validatePreferences(ctx, existingValSet.Preferences, preferences)
+			if err != nil {
+				return types.ValidatorSetPreferences{}, err
+			}
+
+			// (1) no valset and no existing delegation, so store new valset
+			return types.ValidatorSetPreferences{Preferences: validPreference}, nil
+		}
```

Comment:
IMO having a nested error is not a good design. An error should simply mean stopping execution from where the error occurred and no further logic should be based on that error occurring.